compiler/AST/FnSymbol.cpp

```diff
@@ -735,10 +735,44 @@ TagGenericResult FnSymbol::tagIfGeneric(SymbolMap* map, bool abortOK) {
 // 'map' is expected to be non-NULL if this function has been instantiated.
 //
 bool FnSymbol::hasGenericFormals(SymbolMap* map) const {
+  bool anyGeneric = false;
+
+  std::vector<unsigned char> formalIsGeneric;
+
+  int count = 0;
   for_formals(formal, this) {
     bool isGeneric = false;
 
     if (formal->type == dtUnknown && formal->typeExpr != NULL) {
+
+      // If the type expression does not depend on a previous generic formal,
+      // we can resolve it now, even if there was a previous generic formal.
+
+      bool dependsOnPreviousGeneric = false;
+      int i = 0;
+      if (anyGeneric) {
+        for_formals(prevFormal, this) {
+          if (i >= count)
+            break;
+
+          int cursz = formalIsGeneric.size();
+          if (i < cursz && formalIsGeneric[i] == 1) {
```

Comment:
Makes me think that we may as well update `formalIsGeneric` eagerly, rather than lazily, to simplify quite a bit of logic around it. OK to leave that for another time.

Reply:
The patch used to do that, but I got slightly better compiler performance doing it this way. I do it this way in order to avoid allocating the `std::vector`'s buffer in the event the function is entirely concrete.

Reply:
(LLVM's data structures like `SmallVector` are designed for cases like this and I look forward to the day we can use them throughout the compiler).